### ISSUE TITLE
[develop] Not all processes are properly stopped when running test suite

### ISSUE THREAD

**Issue by rallytime** [saltstack/salt#33920]
### Description of Issue/Question
When running the test suite on the `develop` branch, not all of the test processes that are spun up by the integration tests are shut down appropriately when the integration tests finish. The syndic process is always left hanging around, the sub-minion process is often (though not always) left running, and sometimes even the master and minion processes remain.

### Steps to Reproduce Issue
Run any integration test on the develop. Then look at the process list once the test(s) finish executing and CLI returns. Multiple test runs will leave many processes hanging around. This can cause socket errors when starting the test suite again.
```
#root@rallytime:~# ps aux | grep salt
root      7425  0.0  1.7 309728 35716 pts/4    Sl   21:18   0:00 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config -l quiet
root      7426  0.0  1.8 465388 37568 pts/4    Sl   21:18   0:00 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config -l quiet
root      7433  0.5  0.0      0     0 pts/4    Z    21:18   0:01 [cli_salt_master] <defunct>
root      7436  0.6  2.5 334736 52200 pts/4    Sl   21:18   0:02 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config -l quiet
root      7437  0.4  2.3 331072 48472 pts/4    Sl   21:18   0:01 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config -l quiet
root      7684  0.1  1.8 313348 37500 pts/4    Sl   21:18   0:00 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_minion.py -c /tmp/salt-tests-tmpdir/config -l quiet --disable-keepalive
root      7983  0.0  1.8 313348 37484 pts/4    Sl   21:18   0:00 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_minion.py -c /tmp/salt-tests-tmpdir/config/sub-minion -l quiet --disable-keepalive
root      8085  0.5  2.3 548940 47936 pts/4    Sl   21:18   0:01 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_syndic.py -c /tmp/salt-tests-tmpdir/config/syndic-minion -l quiet
root      8400  0.0  1.9 334200 39968 pts/4    Sl   21:18   0:00 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_syndic.py -c /tmp/salt-tests-tmpdir/config/syndic-minion -l quiet
```

### Versions Report
```
root@rallytime:~# salt --versions
Salt Version:
           Salt: 2016.3.0-688-g62edce6

Dependency Versions:
           cffi: 0.8.6
       cherrypy: 5.0.1
       dateutil: 2.4.2
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: 0.21.1
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: 2.10
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.5.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5

System Versions:
           dist: Ubuntu 14.04 trusty
        machine: x86_64
        release: 4.4.0-x86_64-linode63
         system: Linux
        version: Ubuntu 14.04 trusty
```

**Comment by rallytime** [comment]
I tracked this down using git bisect and the cause of this issue is from #33108, specifically the 57fbe717b9b1833a2474bd286ae0c12233b5ebd0 commit.

FYI @s0undt3ch

**Comment by rallytime** [comment]
So I tested this again with the latest develop because I thought #35632 might clear this up. But it seems to be making it worse. I am running off of the current HEAD of the develop branch of both salt and salt-testing. After doing a simple `python tests/runtests -s` to run the shell test subsection, and then looking for running procs, I see this result:
```
# ps aux | grep salt
root      9061  0.8  2.4 405732 48972 pts/1    Sl   16:44   0:03 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_syndic.py -c /tmp/salt-tests-tmpdir/config/syndic-minion -l quiet
root      9160  0.0  1.0 296672 21592 pts/1    S    16:44   0:00 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_syndic.py -c /tmp/salt-tests-tmpdir/config/syndic-minion -l quiet MultiprocessingLoggingQueue
root      9659  0.2  1.9 386392 39504 pts/1    Sl   16:45   0:00 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config -l quiet EventPublisher
root      9661  0.0  1.7 382140 35288 pts/1    Sl   16:45   0:00 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config -l quiet
root      9663  0.9  2.7 407872 56000 pts/1    Sl   16:45   0:03 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config -l quiet Maintenance
root      9664  0.1  2.8 406196 59080 pts/1    Sl   16:45   0:00 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config -l quiet ReqServer_ProcessManager
root      9666  0.0  1.9 537544 38756 pts/1    Sl   16:45   0:00 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config -l quiet MWorkerQueue
root      9673  1.5  3.5 578536 72844 pts/1    Sl   16:45   0:06 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config -l quiet MWorker-0
root      9675  1.1  4.6 599724 95088 pts/1    Sl   16:45   0:04 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config -l quiet MWorker-1
root      9677  1.8  4.9 604896 100216 pts/1   Sl   16:45   0:07 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config -l quiet MWorker-2
root     10431  0.1  1.6 381608 34656 pts/1    Sl   16:45   0:00 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config/syndic-master -l quiet EventPublisher
root     10433  0.0  1.7 382148 35096 pts/1    Sl   16:45   0:00 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config/syndic-master -l quiet
root     10435  0.7  2.7 407856 55700 pts/1    Sl   16:45   0:02 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config/syndic-master -l quiet Maintenance
root     10438  0.0  1.9 537552 38772 pts/1    Sl   16:45   0:00 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config/syndic-master -l quiet MWorkerQueue
root     10445  0.6  3.0 485980 62492 pts/1    Sl   16:45   0:02 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config/syndic-master -l quiet MWorker-0
root     10447  0.6  4.0 500036 82984 pts/1    Sl   16:45   0:02 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config/syndic-master -l quiet MWorker-1
root     10448  0.6  4.0 500004 83036 pts/1    Sl   16:45   0:02 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config/syndic-master -l quiet MWorker-2
root     10453  1.5  3.0 637968 62764 pts/1    Sl   16:45   0:05 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_syndic.py -c /tmp/salt-tests-tmpdir/config/syndic-minion -l quiet MinionProcessManager
root     10954  0.0  2.0 406148 40940 pts/1    Sl   16:46   0:00 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_syndic.py -c /tmp/salt-tests-tmpdir/config/syndic-minion -l quiet MinionProcessManager
root     15099  0.0  1.5 308624 32512 ?        S    16:49   0:00 /usr/bin/python /tmp/salt-testdaemon-bNPq_R/correct_usage/bin/salt-master -d --config-dir /tmp/salt-testdaemon-bNPq_R/correct_usage/etc/salt MultiprocessingLoggingQueue
root     15100  0.0  1.8 317868 37852 ?        Sl   16:49   0:00 /usr/bin/python /tmp/salt-testdaemon-bNPq_R/correct_usage/bin/salt-master -d --config-dir /tmp/salt-testdaemon-bNPq_R/correct_usage/etc/salt ZeroMQPubServerChannel
root     15103  0.0  1.6 309672 33908 ?        S    16:49   0:00 /usr/bin/python /tmp/salt-testdaemon-bNPq_R/correct_usage/bin/salt-master -d --config-dir /tmp/salt-testdaemon-bNPq_R/correct_usage/etc/salt EventPublisher
root     17136  0.0  0.1  11756  2136 pts/1    S+   16:52   0:00 grep --color=auto salt
```

**Comment by s0undt3ch** [comment]
@rallytime does the machine from where you got that `ps` output has `psutils` installed?

**Comment by cachedout** [comment]
How about with this? https://github.com/saltstack/salt/pull/36399

**Comment by rallytime** [comment]
@cachedout I just tested with #36399. I checked out your branch, cleaned out any potential stale .pyc files, and made sure to install the latest psutils. Here's the output:
```
# ./runtests.py -R
/root/SaltStack/salt/salt/utils/psutil_compat.py:56: DeprecationWarning: psutil.network_io_counters is deprecated; use psutil.net_io_counters() instead
  net_io_counters = psutil.network_io_counters()
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Transplanting configuration files to '/tmp/salt-tests-tmpdir/config'
 * Current Directory: /root/SaltStack/salt
 * Test suite is running under PID 10503
 * Logging tests on /tmp/salt-runtests.log
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Max open files settings is too low (soft: 1024, hard: 4096) for running the tests
 * Trying to raise the limits to soft: 3072, hard: 4096
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Setting up Salt daemons to execute tests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Starting salt-master ... STARTED!
 * Starting salt-minion ... /root/SaltStack/salt/salt/utils/psutil_compat.py:56: DeprecationWarning: psutil.network_io_counters is deprecated; use psutil.net_io_counters() instead
  net_io_counters = psutil.network_io_counters()
 * Starting salt-minion ... STARTED!
 * Starting sub salt-minion ... /root/SaltStack/salt/salt/utils/psutil_compat.py:56: DeprecationWarning: psutil.network_io_counters is deprecated; use psutil.net_io_counters() instead
  net_io_counters = psutil.network_io_counters()
 * Starting sub salt-minion ... STARTED!
 * Starting syndic salt-master ... STARTED!
 * Starting salt-syndic ... STARTED!
==================================================================================================================================
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Starting Renderers Tests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.
WARNING: A misbehaving test has modified the working directory!
The test suite has reset the working directory on tearDown() to /root/SaltStack/salt/tests


----------------------------------------------------------------------
Ran 1 test in 4.437s

OK
Traceback (most recent call last):
  File "./runtests.py", line 624, in <module>
    main()
  File "./runtests.py", line 609, in main
    status = parser.run_integration_tests()
  File "./runtests.py", line 549, in run_integration_tests
    status.append(self.run_integration_suite(**TEST_SUITES[suite]))
  File "/root/SaltStack/salt/tests/integration/__init__.py", line 1276, in __exit__
    self.sub_minion_process.terminate()
  File "/root/SaltStack/salt/tests/integration/__init__.py", line 556, in terminate
    terminate_process_pid(self._process.pid)
  File "/root/SaltStack/salt-testing/salttesting/helpers.py", line 1084, in terminate_process_pid
    cmdline = process.cmdline()
TypeError: 'list' object is not callable
Process SignalHandlingMultiprocessingProcess-4:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/root/SaltStack/salt/salt/utils/process.py", line 651, in _run
    return self._original_run()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/root/SaltStack/salt/tests/integration/__init__.py", line 545, in _start
    terminate_process_pid(terminal.pid)
  File "/root/SaltStack/salt-testing/salttesting/helpers.py", line 1084, in terminate_process_pid
    cmdline = process.cmdline()
TypeError: 'list' object is not callable
```
At which point the whole thing hung up until I hit control-c. At that point, the rest of the shutdown looked like this:
```
^CProcess SignalHandlingMultiprocessingProcess-2:
The Salt Minion is shutdown. Minion received a SIGINT. Exited.
Error in atexit._run_exitfuncs:Process SignalHandlingMultiprocessingProcess-6:
The Salt Minion is shutdown. Minion received a SIGINT. Exited.
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/root/SaltStack/salt/salt/utils/process.py", line 651, in _run
Process SignalHandlingMultiprocessingProcess-5:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/root/SaltStack/salt/salt/utils/process.py", line 651, in _run
    return self._original_run()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/root/SaltStack/salt/tests/integration/__init__.py", line 545, in _start
    terminate_process_pid(terminal.pid)
  File "/root/SaltStack/salt-testing/salttesting/helpers.py", line 1084, in terminate_process_pid
    cmdline = process.cmdline()
TypeError: 'list' object is not callable
    return self._original_run()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/root/SaltStack/salt/tests/integration/__init__.py", line 545, in _start
    terminate_process_pid(terminal.pid)
  File "/root/SaltStack/salt-testing/salttesting/helpers.py", line 1084, in terminate_process_pid
    cmdline = process.cmdline()
TypeError: 'list' object is not callable
Process SignalHandlingMultiprocessingProcess-3:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/root/SaltStack/salt/salt/utils/process.py", line 651, in _run
    return self._original_run()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/root/SaltStack/salt/tests/integration/__init__.py", line 545, in _start
    terminate_process_pid(terminal.pid)
  File "/root/SaltStack/salt-testing/salttesting/helpers.py", line 1084, in terminate_process_pid
    cmdline = process.cmdline()
TypeError: 'list' object is not callable
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/root/SaltStack/salt/salt/utils/process.py", line 651, in _run
    return self._original_run()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/root/SaltStack/salt/tests/integration/__init__.py", line 545, in _start
    terminate_process_pid(terminal.pid)
  File "/root/SaltStack/salt-testing/salttesting/helpers.py", line 1084, in terminate_process_pid
    cmdline = process.cmdline()
TypeError: 'list' object is not callable
The Salt Syndic is shutdown. Syndic received a SIGINT. Exited.

Traceback (most recent call last):
  File "/usr/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "/usr/lib/python2.7/multiprocessing/util.py", line 325, in _exit_function
    p.join()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 145, in join
    res = self._popen.wait(timeout)
  File "/usr/lib/python2.7/multiprocessing/forking.py", line 154, in wait
    return self.poll(0)
  File "/usr/lib/python2.7/multiprocessing/forking.py", line 135, in poll
    pid, sts = os.waitpid(self.pid, flag)
KeyboardInterrupt
Error in sys.exitfunc:
The salt master is shutdown. Master received a SIGINT. Exited.
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 47896)
Traceback (most recent call last):
  File "/usr/lib/python2.7/SocketServer.py", line 593, in process_request_thread
----------------------------------------
 ----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 47800)
 ----------------------------------------
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 47874)
Traceback (most recent call last):
Exception happened during processing of request from ('127.0.0.1', 47832)
Exception happened during processing of request from ('127.0.0.1', 47816)
  File "/usr/lib/python2.7/SocketServer.py", line 593, in process_request_thread
    self.finish_request(request, client_address)
  File "/usr/lib/python2.7/SocketServer.py", line 334, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib/python2.7/SocketServer.py", line 649, in __init__
    self.handle()
  File "/root/SaltStack/salt/tests/integration/__init__.py", line 404, in handle
    log.exception(exc)
  File "/usr/lib/python2.7/logging/__init__.py", line 1185, in exception
    self.error(msg, *args, **kwargs)
  File "/usr/lib/python2.7/logging/__init__.py", line 1178, in error
    self._log(ERROR, msg, args, **kwargs)
  File "/root/SaltStack/salt/salt/log/setup.py", line 309, in _log
    self.finish_request(request, client_address)
Traceback (most recent call last):
Traceback (most recent call last):
Traceback (most recent call last):
root@rallytime:~/SaltStack/salt/tests# The salt master is shutdown. Master received a SIGINT. Exited.
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/usr/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "/usr/lib/python2.7/multiprocessing/util.py", line 321, in _exit_function
    p._popen.terminate()
  File "/usr/lib/python2.7/multiprocessing/forking.py", line 171, in terminate
    os.kill(self.pid, signal.SIGTERM)
OSError: [Errno 3] No such process
Error in sys.exitfunc:
Traceback (most recent call last):
  File "/usr/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "/usr/lib/python2.7/multiprocessing/util.py", line 321, in _exit_function
    p._popen.terminate()
  File "/usr/lib/python2.7/multiprocessing/forking.py", line 171, in terminate
    os.kill(self.pid, signal.SIGTERM)
OSError: [Errno 3] No such process
^C
```
Luckily, there aren't any processes hanging around!
```
# ps aux | grep salt
root     11845  0.0  0.1  11752  2160 pts/3    S+   21:33   0:00 grep --color=auto salt
```
But that hanging behavior isn't exactly desired, either.

**Comment by rallytime** [comment]
More information: After running a couple more test runs right after submitting my last comment (nothing changed on my system, just figured I would test a couple other areas), if I hit control-c more than once and too rapidly, I have hanging procs still:
```
# ps aux | grep salt
root     12527  0.6  2.9 397260 60604 pts/3    Sl   21:34   0:01 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config/syndic-master -l quiet Master_ProcessManager
root     12575  0.7  2.5 402220 52460 pts/3    Sl   21:34   0:01 /usr/bin/python /tmp/salt-tests-tmpdir/scripts/cli_salt_master.py -c /tmp/salt-tests-tmpdir/config/syndic-master -l quiet MWorker-0
```

**Comment by rallytime** [comment]
I just tested this again and it appears to be resolved. Possibly fixed by #36745 in addition to #36399.
